src/aurora/gff3writer.cpp

```diff
@@ -0,0 +1,491 @@
+	}
+}
+
+size_t GFF3Writer::createField(GFF3Struct::FieldType type, const Common::UString &label) {
+	// Create a field index.
+	size_t index = _fields.size();
+
+	// Create field.
+	GFF3Writer::Field field;
+	field.type = type;
+	field.labelIndex = addLabel(label);
+	_fields.push_back(field);
+
+	return index;
+}
+
+GFF3WriterStructPtr GFF3WriterList::addStruct(const Common::UString &label) {
+	// Create the structure pointer.
+	GFF3WriterStructPtr strct(
+			boost::make_shared<GFF3WriterStruct>(_parent, static_cast<uint32>(_parent->_structs.size()) - 1));
+
+	// Create a field index.
+	_strcts.push_back(_parent->_structs.size());
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeStruct;
+	field.uint32Value = static_cast<uint32>(_parent->_structs.size());
+
+	// Add the label.
+	field.labelIndex = _parent->addLabel(label);
+
+	// Insert the newly created struct into the struct vector.
+	_parent->_structs.push_back(strct);
+
+	// Insert the struct to the field vector.
+	_parent->_fields.push_back(field);
+
+	return strct;
+}
+
+size_t GFF3WriterList::getSize() const {
+	return _strcts.size();
+}
+
+GFF3WriterList::GFF3WriterList(GFF3Writer *parent) : _parent(parent) {
+}
+
+uint32 GFF3WriterStruct::getId() const {
+	return _id;
+}
+
+size_t GFF3WriterStruct::getFieldCount() const {
+	return _fieldIndices.size();
+}
+
+GFF3WriterStructPtr GFF3WriterStruct::addStruct(const Common::UString &label) {
+	// Create the structure pointer.
+	GFF3WriterStructPtr strct(
+			boost::make_shared<GFF3WriterStruct>(_parent, static_cast<uint32>(_parent->_structs.size()) - 1));
+
+	// Create a field index.
+	_fieldIndices.push_back(_parent->_fields.size());
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeStruct;
+	field.uint32Value = static_cast<uint32>(_parent->_structs.size());
+
+	// Add the label.
+	field.labelIndex = _parent->addLabel(label);
+
+	// Insert the newly created struct into the struct vector.
+	_parent->_structs.push_back(strct);
+
+	// Insert the struct to the field vector.
+	_parent->_fields.push_back(field);
+	
+	return strct;
+}
+
+GFF3WriterListPtr GFF3WriterStruct::addList(const Common::UString &label) {
+	// Create the list pointer.
+	GFF3WriterListPtr strct(boost::make_shared<GFF3WriterList>(_parent));
+
+	// Create a field index.
+	_fieldIndices.push_back(_parent->_fields.size());
+	GFF3Writer::Field field;
+	field.type = GFF3Struct::kFieldTypeList;
+	field.uint32Value = static_cast<uint32>(_parent->_lists.size());
+
+	// Add the label.
+	field.labelIndex = _parent->addLabel(label);
+
+	// Insert the newly created list into the lists vector.
+	_parent->_lists.push_back(strct);
+
+	// Insert the list to the field vector.
+	_parent->_fields.push_back(field);
+
+	return strct;
+}
+
+void GFF3WriterStruct::addByte(const Common::UString &label, byte value) {
+	size_t index = _parent->createField(GFF3Struct::kFieldTypeByte, label);
+	_parent->_fields[index].uint32Value = value;
+	_fieldIndices.push_back(index);
```

Comment:
I'd go even further. `GFF3Writer::Field &GFF3WriterStruct::createField(GFF3Struct::FieldType type, const Common::UString &label)`. Calls `_parent->createField()`, put the index into `_fieldIndices` and returns a direct reference to the field.

Then you only need to do
```
createField(GFF3Struct::kFieldTypeByte, label).uint32Value = value;
```
etc. in each method.